ec2_asg: various fixes

tremble opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_asg

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
```
props doesn't always have an instances key

during 'replace' return changed=False when there are no instances to change.

This primarily occurs when spinning up a new ASG, and avoids people needing
separate plays for creating the ASG and updating the ASG to make changes
```

tremble committed:
ec2_asg: props doesn't always have an instances key
during 'replace' return changed=False when there are no instances to change.

This primarily occurs when spinning up a new ASG, and avoids people needing
separate plays for creating the ASG and updating the ASG and making changes

tremble commented:
Specifically avoids the following stacktrace:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'instances'
fatal: [OpenPlatformIntNode]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_eVTWmk/ansible_module_ec2_asg.py\", line 877, in <module>\n    main()\n  File \"/tmp/ansible_eVTWmk/ansible_module_ec2_asg.py\", line 871, in main\n    replace_changed, asg_properties=replace(connection, module)\n  File \"/tmp/ansible_eVTWmk/ansible_module_ec2_asg.py\", line 603, in replace\n    new_instances, old_instances = get_instances_by_lc(props, lc_check, instances)\n  File \"/tmp/ansible_eVTWmk/ansible_module_ec2_asg.py\", line 667, in get_instances_by_lc\n    for i in props['instances']:\nKeyError: 'instances'\n", "module_stdout": "", "msg": "MODULE FAILURE"}

```

gregdek commented:
Thanks @tremble for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

tremble committed:
ec2_asg: as_group.min_size can respond with None rather than 0, set it to 0 to avoid type errors

tremble committed:
ec2_asg: Make sure we have finished scaling up the ASG before continuing

When min_size is 0 and desired_size == batch_size we can end up killing
all nodes before the new ones are ready.

tremble committed:
ec2_asg: desired_capacity is an optional argument

pass it into terminate_batch because we can't guarantee a valid value in the
parameters

tremble committed:
ec2_asg: Use the current desired capacity as the basis for how many nodes to
wait for, rather than the minimum number of nodes

If the group is using a metric to autoscale try to maintain a consistent number of viable
instances *after* nodes have been terminated.  This behaviour is consistent with CloudFormation
behaviour.

tremble committed:
ec2_asg: Treat nodes in intermediate states as Pending or Terminating as appropriate

tremble committed:
ec2_asg: Only work through a list of old instances when performing replacements

tremble commented:
Includes a fix for #5555
